concretizer: unify logic for spec conditionals

tgamblin opened the pull request:
Depends on #20638, which should go in first.

This builds on #20638 by unifying all the places in the concretizer where things are conditional on specs.  Previously, we duplicated a common spec conditional pattern for dependencies, virtual providers, conflicts, and externals.  That was introduced in #20423 and refined in #20507, and roughly looked as follows.

Given some directives in a package like:

```python
depends_on("foo@1.0+bar", when="@2.0+variant")
provides("mpi@2:", when="@1.9:")
```

We handled the `@2.0+variant` and `@1.9:` parts by generating generated `dependency_condition()`, `required_dependency_condition()`, and `imposed_dependency_condition()` facts to trigger rules like this:

```prolog
dependency_conditions_hold(ID, Parent, Dependency) :-
  attr(Name, Arg1)             : required_dependency_condition(ID, Name, Arg1);
  attr(Name, Arg1, Arg2)       : required_dependency_condition(ID, Name, Arg1, Arg2);
  attr(Name, Arg1, Arg2, Arg3) : required_dependency_condition(ID, Name, Arg1, Arg2, Arg3);
  dependency_condition(ID, Parent, Dependency);
  node(Parent).
```

And we handled `foo@1.0+bar` and `mpi@2:` parts ("imposed constraints") like this:

```prolog
attr(Name, Arg1, Arg2) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2).

attr(Name, Arg1, Arg2, Arg3) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2, Arg3).
```

These rules were repeated with different input predicates for requirements (e.g., `required_dependency_condition`) and imposed constraints (e.g., `imposed_dependency_condition`) throughout `concretize.lp`. In #20638 it got to be a bit confusing, because we used the same `dependency_condition_holds` predicate to impose constraints on conditional dependencies and virtual providers. So, even though the pattern was repeated, some of the conditional rules were conjoined in a weird way.

Instead of repeating this pattern everywhere, we now have *one* set of consolidated rules for conditions:

```prolog
condition_holds(ID) :-
  condition(ID);
  attr(Name, A1)         : condition_requirement(ID, Name, A1);
  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).

attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
attr(Name, A1, A2)     :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2).
attr(Name, A1, A2, A3) :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2, A3).
```

this allows us to use `condition(ID)` and `condition_holds(ID)` to encapsulate the conditional logic on specs in all the scenarios where we need it.  Instead of defining predicates for the requirements and imposed constraints, we generate the condition inputs with generic facts, and define predicates to associate the condition ID with a particular scenario.  So, now, the generated facts for a condition look like this:

```prolog
condition(121).                                                                          
condition_requirement(121,"node","cairo").                                               
condition_requirement(121,"variant_value","cairo","fc","True").                          
imposed_constraint(121,"version_satisfies","fontconfig","2.10.91:").                     
dependency_condition(121,"cairo","fontconfig").                                          
dependency_type(121,"build").                                                            
dependency_type(121,"link").  
```

The requirements and imposed constraints are generic, and we associate them with their meaning via the id.  Here, `dependency_condition(121,  "cairo", "fontconfig")` tells us that condition 121 has to do with the dependency of `cairo` on `fontconfig`, and the conditional dependency rules just become:

```prolog
dependency_holds(Package, Dependency, Type) :-
  dependency_condition(ID, Package, Dependency),
  dependency_type(ID, Type),
  condition_holds(ID).
```

Dependencies, virtuals, conflicts, and externals all now use similar patterns, and the logic for generating condition facts is common to all of them on the python side, as well.  The more specific routines like `package_dependencies_rules` just call `self.condition(...)` to get an id and generate requirements and imposed constraints, then they generate their extra facts wth the returned id, like this:

```python
    def package_dependencies_rules(self, pkg, tests):
        """Translate 'depends_on' directives into ASP logic."""
        for _, conditions in sorted(pkg.dependencies.items()):
            for cond, dep in sorted(conditions.items()):
                condition_id = self.condition(cond, dep.spec, pkg.name)  # create a condition and get its id
                self.gen.fact(fn.dependency_condition(  # associate specifics about the dependency w/the id
                    condition_id, pkg.name, dep.spec.name
                ))
        # etc.
```

- [x] unify generation and logic for conditions
- [x] use unified logic for dependencies
- [x] use unified logic for virtuals
- [x] use unified logic for conflicts
- [x] use unified logic for externals

tgamblin committed:
concretizer: unify logic for spec conditionals

This builds on #20638 by unifying all the places in the concretizer where
things are conditional on specs. Previously, we duplicated a common spec
conditional pattern for dependencies, virtual providers, conflicts, and
externals. That was introduced in #20423 and refined in #20507, and
roughly looked as follows.

Given some directives in a package like:

```python
depends_on("foo@1.0+bar", when="@2.0+variant")
provides("mpi@2:", when="@1.9:")
```

We handled the `@2.0+variant` and `@1.9:` parts by generating generated
`dependency_condition()`, `required_dependency_condition()`, and
`imposed_dependency_condition()` facts to trigger rules like this:

```prolog
dependency_conditions_hold(ID, Parent, Dependency) :-
  attr(Name, Arg1)             : required_dependency_condition(ID, Name, Arg1);
  attr(Name, Arg1, Arg2)       : required_dependency_condition(ID, Name, Arg1, Arg2);
  attr(Name, Arg1, Arg2, Arg3) : required_dependency_condition(ID, Name, Arg1, Arg2, Arg3);
  dependency_condition(ID, Parent, Dependency);
  node(Parent).
```

And we handled `foo@1.0+bar` and `mpi@2:` parts ("imposed constraints")
like this:

```prolog
attr(Name, Arg1, Arg2) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2).

attr(Name, Arg1, Arg2, Arg3) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2, Arg3).
```

These rules were repeated with different input predicates for
requirements (e.g., `required_dependency_condition`) and imposed
constraints (e.g., `imposed_dependency_condition`) throughout
`concretize.lp`. In #20638 it got to be a bit confusing, because we used
the same `dependency_condition_holds` predicate to impose constraints on
conditional dependencies and virtual providers. So, even though the
pattern was repeated, some of the conditional rules were conjoined in a
weird way.

Instead of repeating this pattern everywhere, we now have *one* set of
consolidated rules for conditions:

```prolog
condition_holds(ID) :-
  condition(ID);
  attr(Name, A1)         : condition_requirement(ID, Name, A1);
  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).

attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
attr(Name, A1, A2)     :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2).
attr(Name, A1, A2, A3) :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2, A3).
```

this allows us to use `condition(ID)` and `condition_holds(ID)` to
encapsulate the conditional logic on specs in all the scenarios where we
need it. Instead of defining predicates for the requirements and imposed
constraints, we generate the condition inputs with generic facts, and
define predicates to associate the condition ID with a particular
scenario. So, now, the generated facts for a condition look like this:

```prolog
condition(121).
condition_requirement(121,"node","cairo").
condition_requirement(121,"variant_value","cairo","fc","True").
imposed_constraint(121,"version_satisfies","fontconfig","2.10.91:").
dependency_condition(121,"cairo","fontconfig").
dependency_type(121,"build").
dependency_type(121,"link").
```

The requirements and imposed constraints are generic, and we associate
them with their meaning via the id. Here, `dependency_condition(121,
"cairo", "fontconfig")` tells us that condition 121 has to do with the
dependency of `cairo` on `fontconfig`, and the conditional dependency
rules just become:

```prolog
dependency_holds(Package, Dependency, Type) :-
  dependency_condition(ID, Package, Dependency),
  dependency_type(ID, Type),
  condition_holds(ID).
```

Dependencies, virtuals, conflicts, and externals all now use similar
patterns, and the logic for generating condition facts is common to all
of them on the python side, as well. The more specific routines like
`package_dependencies_rules` just call `self.condition(...)` to get an id
and generate requirements and imposed constraints, then they generate
their extra facts with the returned id, like this:

```python
    def package_dependencies_rules(self, pkg, tests):
        """Translate 'depends_on' directives into ASP logic."""
        for _, conditions in sorted(pkg.dependencies.items()):
            for cond, dep in sorted(conditions.items()):
                condition_id = self.condition(cond, dep.spec, pkg.name)  # create a condition and get its id
                self.gen.fact(fn.dependency_condition(  # associate specifics about the dependency w/the id
                    condition_id, pkg.name, dep.spec.name
                ))
        # etc.
```

- [x] unify generation and logic for conditions
- [x] use unified logic for dependencies
- [x] use unified logic for virtuals
- [x] use unified logic for conflicts
- [x] use unified logic for externals

LocalWords:  concretizer mpi attr Arg concretize lp cairo fc fontconfig
LocalWords:  virtuals def pkg cond dep fn refactor github py

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -307,29 +289,25 @@ real_node(Package) :- node(Package), not external(Package).
+    node(Package).
+
+external_conditions_hold(Package, LocalIndex) :-
+    possible_external(ID, Package, LocalIndex), condition_holds(ID).
```
We need to add `possible_external` among the `#defined` atoms.

tgamblin replied:
done.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -307,29 +289,25 @@ real_node(Package) :- node(Package), not external(Package).
 
 % if an external version is selected, the package is external and
 % we are using the corresponding spec
+% TODO: this only checks that the *version* is the same, but not that
+% TODO: other attributes are. Consider modeling externals with imposed
+% TODO: constraints (to ensure they all match), not by writing out the
+% TODO: match criteria like this (we'll always forget something)
```
This is legacy code stemming from the previous formulation of rules for externals. Everything is checked, but it's indirect and definitely not clear (we'll use the external version weight for external `ID` only if external `ID` is the one selected). I'll check if we can write this rule in a way it's less cryptic.

alalazo replied:
This should be a better substitute for the rule below the comment:
```
% a package is external if we are using an external spec for it
external(Package) :- external_spec_selected(Package, _).

% we can't use the weight for an external version if we don't use the 
% corresponding external spec.
:- version(Package, Version),
   version_weight(Package, Weight),
   external_version_declared(Package, Version, Weight, ID),
   not external(Package).
```
(or at least it seems clearer to me).

tgamblin replied:
ok this is addressed.

alalazo reviewed (commented):
LGTM. One minor comment on having a term `#defined`.

tgamblin commented:
@alalazo: this is ready for another look.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -36,16 +36,54 @@ version_satisfies(Package, Constraint)
+%-----------------------------------------------------------------------------
+% conditions are specified with `condition_requirement` and hold when
+% corresponding spec attributes hold.
+condition_holds(ID) :-
+  condition(ID);
+  attr(Name, A1)         : condition_requirement(ID, Name, A1);
+  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
+  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).
+
+% conditions that hold impose constraints on other specs
+attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
+attr(Name, A1, A2)     :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2).
+attr(Name, A1, A2, A3) :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2, A3).
```
The tests pass for me locally with the following diff:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 1dfdf40cc3..3a643f9215 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -50,9 +50,22 @@ version_satisfies(Package, Constraint)
 % corresponding spec attributes hold.
 condition_holds(ID) :-
   condition(ID);
-  attr(Name, A1)         : condition_requirement(ID, Name, A1);
-  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
-  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).
+  attr(Name, A1)              : condition_requirement(ID, Name, A1);
+  attr(Name, A1, A2)          : condition_requirement(ID, Name, A1, A2);
+  attr(Name, A1, A2, A3)      : condition_requirement(ID, Name, A1, A2, A3);
+  % If this condition is a dependency, we need to have at least
+  % a dependency type defined
+  dependency_type_defined(ID) : dependency_condition(ID, _, _);
+  % Since a condition that holds on a dependency triggers the presence
+  % of the dependency, make it such that if the dependent is an external,
+  % then the condition doesn't hold
+  #false                      : dependency_of_an_external(ID).
+
+% Ensure we have at least a dependency type for dependency ID
+dependency_type_defined(ID) :- dependency_type(ID, _).
+
+% True if ID is a dependency condition and the dependent is an external
+dependency_of_an_external(ID) :- dependency_condition(ID, Package, _), external(Package).
 
 % conditions that hold impose constraints on other specs
 attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
```

tgamblin replied:
Taking a look at this -- I want to see if I can find a way to consolidate the logic here a bit more.  This sort of pollutes the general conditions with information about externals, but maybe there is no way around it...

alalazo reviewed (changes requested):
This basically LGTM. I think we should merge the PR immediately after tests are passing, but we may reconsider adding it to v0.16.1 since the refactor doesn't directly impact users and may introduce bugs that are not caught by the current test suite.

tgamblin commented:
> we may reconsider adding it to v0.16.1 since the refactor doesn't directly impact users and may introduce bugs that are not caught by the current test suite.

I moved this to `0.16.2` for now.

tgamblin committed:
bugfix: do not generate dep conditions when no dependency

We only consider test dependencies some of the time. Some packages are
*only* test dependencies. Spack's algorithm was previously generating
dependency conditions that could hold, *even* if there was no potential
dependency type.

- [x] change asp.py so that this can't happen -- we now only generate
      dependency types for possible dependencies.

tgamblin committed:
bugfix: allow imposed constraints to be overridden in special cases

In most cases, we want condition_holds(ID) to imply any imposed
constraints associated with the ID. However, the dependency relationship
in Spack is special because it's "extra" conditional -- a dependency
*condition* may hold, but we have decided that externals will not have
dependencies, so we need a way to avoid having imposed constraints appear
for nodes that don't exist.

This introduces a new rule that says that constraints are imposed
*unless* we define `do_not_impose(ID)`. This allows rules like
dependencies, which rely on more than just spec conditions, to cancel
imposed constraints.

We add one special case for this: dependencies of externals.

tgamblin commented:
@alalazo: this should be ready for re-review now.  Sorry it took so long :).

See what you think of the last commit I pushed.  It's actually the only one necessary to ensure that the tests pass -- though I think having the bugfix for conditions with no dependency is still cleaner.

The issue (that you diagnosed above) is that in my model, `condition_holds(ID)` always implies imposed constraints, but it shouldn't.  Not all conditions are *just* dependent on spec properties; they are also dependent on things like whether a package is an external.  So I wanted a way to attach extra logic like that to conditions *without* having to lump it all into the same rule.

So, now you can derive `do_not_impose(ID)` to cancel the imposed constraints:

```prolog
% condition_holds(ID) implies all imposed_constraints, unless something no.
% this allows imposed constraints to be canceled in special cases.
impose(ID) :- condition_holds(ID), not do_not_impose(ID).

attr(Name, A1)         :- impose(ID), imposed_constraint(ID, Name, A1).
attr(Name, A1, A2)     :- impose(ID), imposed_constraint(ID, Name, A1, A2).
attr(Name, A1, A2, A3) :- impose(ID), imposed_constraint(ID, Name, A1, A2, A3).
```

They'll be derived by default (which is what I think we usually want), but you can add exceptions, and the exception logic can go where it most makes sense, e.g. we can now put the logic for this external exception *with* the logic for dependencies:

```prolog
% We cut off dependencies of externals (as we don't really know them).
% Don't impose constraints on dependencies that don't exist.
do_not_impose(ID) :-
  not dependency_holds(Package, Dependency, _),
  dependency_condition(ID, Package, Dependency).
```

I think this is cleaner and more intuitive.  What do you think?

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -54,10 +54,14 @@ condition_holds(ID) :-
   attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
   attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).
 
+% condition_holds(ID) implies all imposed_constraints, unless something no.
```
Is the comment truncated?

tgamblin replied:
thanks! fixed.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -77,15 +81,21 @@ depends_on(Package, Dependency) :- depends_on(Package, Dependency, _).
+  condition_holds(ID),
+  not external(Package).
+
+% We cut off dependencies of externals (as we don't really know them).
+% Don't impose constraints on dependencies that don't exist.
```
Is one of the two lines redundant? (They are both correct, but they seem to say the same thing to me)

tgamblin replied:
IMO they're not redundant -- one's about how spack works and the other's about what we have to do in the solve :)

alalazo reviewed (dismissed):
Just very minor comments. LGTM

alalazo commented:
> So I wanted a way to attach extra logic like that to conditions without having to lump it all into the same rule.

I fully agree with this and I like the flexibility it gives to cancel imposed constraints by either inferring a `do_not_impose` or emitting it as a fact if need be.
